File: InterfacePython/example_sparse_grids_08.py
##############################################################################################################################################################################
# Copyright (c) 2017, Miroslav Stoyanov
#
# This file is part of
# Toolkit for Adaptive Stochastic Modeling And Non-Intrusive ApproximatioN: TASMANIAN
#
# Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions
#    and the following disclaimer in the documentation and/or other materials provided with the distribution.
#
# 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse
#    or promote products derived from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
# INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
# OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
# OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# UT-BATTELLE, LLC AND THE UNITED STATES GOVERNMENT MAKE NO REPRESENTATIONS AND DISCLAIM ALL WARRANTIES, BOTH EXPRESSED AND IMPLIED.
# THERE ARE NO EXPRESS OR IMPLIED WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE, OR THAT THE USE OF THE SOFTWARE WILL NOT INFRINGE ANY PATENT,
# COPYRIGHT, TRADEMARK, OR OTHER PROPRIETARY RIGHTS, OR THAT THE SOFTWARE WILL ACCOMPLISH THE INTENDED RESULTS OR THAT THE SOFTWARE OR ITS USE WILL NOT RESULT IN INJURY OR DAMAGE.
# THE USER ASSUMES RESPONSIBILITY FOR ALL LIABILITIES, PENALTIES, FINES, CLAIMS, CAUSES OF ACTION, AND COSTS AND EXPENSES, CAUSED BY, RESULTING FROM OR ARISING OUT OF,
# IN WHOLE OR IN PART THE USE, STORAGE OR DISPOSAL OF THE SOFTWARE.
##############################################################################################################################################################################

import numpy as np
import Tasmanian

def example_08():

    print("\n---------------------------------------------------------------------------------------------------\n")
    print("Example 8: interpolate different functions demonstrating the different")
    print("           local polynomial rules\n")

    iNumInputs = 2 # using two inputs for testing

    # test the error on a uniform dense grid with 10K points
    iTestGridSize = 100
    dx = np.linspace(-1.0, 1.0, iTestGridSize) # sample on a uniform grid
    aMeshX, aMeshY = np.meshgrid(dx, dx)
    aTestPoints = np.column_stack([aMeshX.reshape((iTestGridSize**2, 1)),
                                   aMeshY.reshape((iTestGridSize**2, 1))])

    def get_error(grid, model, aTestPoints):
        aGridResult = grid.evaluateBatch(aTestPoints)
        aModelResult = np.empty((aTestPoints.shape[0], 1), np.float64)
        for i in range(aTestPoints.shape[0]):
            aModelResult[i,:] = model(aTestPoints[i,:])
        return np.max(np.abs(aModelResult[:,0] - aGridResult[:,0]))

    def smooth_model(aX):
        return np.ones((1,)) * np.exp(-aX[0]**2) * np.cos(aX[1])

    iOrder = 2
    grid_localp     = Tasmanian.makeLocalPolynomialGrid(iNumInputs, 1, 7, iOrder, "localp")
    grid_semilocalp = Tasmanian.makeLocalPolynomialGrid(iNumInputs, 1, 7, iOrder, "semi-localp")

    Tasmanian.loadNeededValues(lambda x, tid : smooth_model(x), grid_localp, 4)
    Tasmanian.loadNeededValues(lambda x, tid : smooth_model(x), grid_semilocalp, 4)

    print("Using smooth model: f(x, y) = exp(-x*x) * cos(y)")
    print(" rule_localp,     points = {0:1d}   error = {1:1.4e}".format(
        grid_localp.getNumPoints(), get_error(grid_localp, smooth_model, aTestPoints)))
    print(" rule_semilocalp, points = {0:1d}   error = {1:1.4e}".format(
        grid_semilocalp.getNumPoints(), get_error(grid_semilocalp, smooth_model, aTestPoints)))
    print(" If the model is smooth, rule_semilocalp has an advantage.\n")

    def zero_model(aX):
        return np.ones((1,)) * np.cos(0.5 * np.pi * aX[0]) * np.cos(0.5 * np.pi * aX[1])

    grid_localp0 = Tasmanian.makeLocalPolynomialGrid(iNumInputs, 1, 6, iOrder, "localp-zero")

    Tasmanian.reloadLoadedPoints(lambda x, tid : zero_model(x), grid_localp, 4)
    Tasmanian.loadNeededValues(lambda x, tid : zero_model(x), grid_localp0, 4)

    print("Using homogeneous model: f(x, y) = cos(pi * x / 2) * cos(pi * y / 2)")
    print(" rule_localp,  points = {0:1d}   error = {1:1.4e}".format(
        grid_localp.getNumPoints(), get_error(grid_localp, zero_model, aTestPoints)))
    print(" rule_localp0, points = {0:1d}   error = {1:1.4e}".format(
        grid_localp0.getNumPoints(), get_error(grid_localp0, zero_model, aTestPoints)))
    print(" The rule_localp0 uses basis tuned for models with zero boundary.")


if (__name__ == "__main__"):
    example_08()

File: SparseGrids/Examples/example_sparse_grids_08.cpp

#include "Tasmanian.hpp"
#include <random>

using namespace std;

/*!
 * \internal
 * \file example_sparse_grids_08.cpp
 * \brief Examples for the Tasmanian Sparse Grid module.
 * \author Miroslav Stoyanov
 * \ingroup TasmanianSGExamples
 *
 * Tasmanian Sparse Grids Example 8.
 * \endinternal
 */

/*!
 * \ingroup TasmanianSGExamples
 * \addtogroup TasmanianSGExamples8 Tasmanian Sparse Grids module, example 8
 *
 * \par Example 8
 * Different local polynomial rules.
 */

/*!
 * \ingroup TasmanianSGExamples8
 * \brief Sparse Grids Example 8: local polynomial rules
 *
 * Local polynomial grids use a hierarchy of basis functions with decreasing support,
 * Tasmanian offers several different types of basis each tuned to different types
 * of models. This example demonstrates the efficiency (i.e., error per number of points)
 * of different local polynomial grids when applied to different models.
 *
 * \snippet SparseGrids/Examples/example_sparse_grids_08.cpp SG_Example_08 example
 */
void sparse_grids_example_08(){
#ifndef __TASMANIAN_DOXYGEN_SKIP
//! [SG_Example_08 example]
#endif

    cout << "\n---------------------------------------------------------------------------------------------------\n";
    cout << std::scientific; cout.precision(4);
    cout << "Example 8: interpolate different functions demonstrating the different\n"
         << "           local polynomial rules\n\n";

    int const num_inputs = 2; // using two inputs for all tests

    // using random points to test the error
    int const num_test_points = 1000;
    std::vector<double> test_points(num_test_points * num_inputs);
    std::minstd_rand park_miller(42);
    std::uniform_real_distribution<double> domain(-1.0, 1.0);
    for(auto &t : test_points) t = domain(park_miller);

    // computes the error between the gird surrogate model and the actual model
    // using the test points, finds the largest absolute error
    auto get_error = [&](TasGrid::TasmanianSparseGrid const &grid,
                         std::function<void(double const x[], double y[], size_t)> model)->
        double{
            std::vector<double> grid_result;
            grid.evaluateBatch(test_points, grid_result);
            double err = 0.0;
            for(int i=0; i<num_test_points; i++){
                double model_result; // using only one output
                model(&test_points[i*num_inputs], &model_result, 0);
                err = std::max(err, std::abs(grid_result[i] - model_result));
            }
            return err;
        };

    // test 1: using smooth function
    auto smooth_model = [](double const x[], double y[], size_t)->
        void{ y[0] = std::exp(-x[0] * x[0]) * std::cos(x[1]); };

    int const order = 2; // at order 1 localp and semilocalp are identical
    auto grid_localp = TasGrid::makeLocalPolynomialGrid(num_inputs, 1, 7, order,
                                                        TasGrid::rule_localp);
    auto grid_semilocalp = TasGrid::makeLocalPolynomialGrid(num_inputs, 1, 7, order,
                                                            TasGrid::rule_semilocalp);

    TasGrid::loadNeededValues(smooth_model, grid_localp, 4);
    TasGrid::loadNeededValues(smooth_model, grid_semilocalp, 4);

    cout << "Using smooth model: f(x, y) = exp(-x*x) * cos(y)\n"
         << " rule_localp,     points = " << grid_localp.getNumPoints()
         << "   error = " << get_error(grid_localp, smooth_model) << "\n"
         << " rule_semilocalp, points = " << grid_semilocalp.getNumPoints()
         << "   error = " << get_error(grid_semilocalp, smooth_model) << "\n"
         << " If the model is smooth, rule_semilocalp has an advantage.\n\n";

    // test 1: using model with zero-boundary conditions
    constexpr double pi = 3.14159265358979323846;
    auto zero_model = [=](double const x[], double y[], size_t)->
        void{ y[0] = std::cos(0.5 * pi * x[0]) * std::cos(0.5 * pi * x[1]); };

    auto grid_localp0 = TasGrid::makeLocalPolynomialGrid(num_inputs, 1, 6, order,
                                                        TasGrid::rule_localp0);

    // the true value indicates overwrite of the currently loaded model
    TasGrid::loadNeededValues<TasGrid::mode_parallel, true>(zero_model, grid_localp, 4);
    TasGrid::loadNeededValues(zero_model, grid_localp0, 4);

    cout << "Using homogeneous model: f(x, y) = cos(pi * x / 2) * cos(pi * y / 2)\n"
         << " rule_localp,   points = " << grid_localp.getNumPoints()
         << "   error = " << get_error(grid_localp, zero_model) << "\n"
         << " rule_localp0,  points = " << grid_localp0.getNumPoints()
         << "   error = " << get_error(grid_localp0, zero_model) << "\n"
         << " The rule_localp0 uses basis tuned for models with zero boundary.\n";

#ifndef __TASMANIAN_DOXYGEN_SKIP
//! [SG_Example_08 example]
#endif
}
